azuredevops/internal/service/feed/resource_feed_permission.go

```diff
@@ -0,0 +1,254 @@
+		FeedId:  &feed_id,
+		Project: &project_id,
+		FeedPermission: &[]feed.FeedPermission{
+			{
+				DisplayName:        &display_name,
+				IdentityDescriptor: identity_response.Descriptor,
+				IdentityId:         identity_response.Id,
+				Role:               &role,
+			},
+		},
+	})
+
+	if err != nil {
+		return fmt.Errorf("updating feed Permission for Feed : %s and Identity : %s, Error: %+v", feed_id, identity_descriptor, err)
+	}
+
+	return resourceFeedPermissionRead(d, m)
+}
+
+func resourceFeedPermissionDelete(d *schema.ResourceData, m interface{}) error {
+	clients := m.(*client.AggregatedClient)
+	feed_id := d.Get("feed_id").(string)
+	identity_descriptor := d.Get("identity_descriptor").(string)
+	role := feed.FeedRoleValues.None
+	project_id := d.Get("project_id").(string)
+
+	identity_response, err := getIdentity(d, m)
+
+	if err != nil {
+		return fmt.Errorf("deleting feed Permission for Feed : %s and Identity : %s, Error: %+v", feed_id, identity_descriptor, err)
+	}
+
+	_, err = clients.FeedClient.SetFeedPermissions(clients.Ctx, feed.SetFeedPermissionsArgs{
+		FeedId:  &feed_id,
+		Project: &project_id,
+		FeedPermission: &[]feed.FeedPermission{
+			{
+				IdentityDescriptor: identity_response.Descriptor,
+				Role:               &role,
+			},
+		},
+	})
+
+	if err != nil {
+		return fmt.Errorf("deleting feed Permission for Feed : %s and Identity : %s, Error: %+v", feed_id, identity_descriptor, err)
+	}
+
+	d.SetId("")
+	return nil
+}
+
+func getIdentity(d *schema.ResourceData, m interface{}) (*identity.Identity, error) {
+	clients := m.(*client.AggregatedClient)
+	identity_descriptor := d.Get("identity_descriptor").(string)
+
+	storageKey, err := clients.GraphClient.GetStorageKey(clients.Ctx, graph.GetStorageKeyArgs{
+		SubjectDescriptor: &identity_descriptor,
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	response, err := clients.IdentityClient.ReadIdentity(clients.Ctx, identity.ReadIdentityArgs{
+		IdentityId: converter.String((*storageKey.Value).String()),
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return response, nil
+}
+
+func getFeedPermission(d *schema.ResourceData, m interface{}) (*feed.FeedPermission, *identity.Identity, error) {
+	clients := m.(*client.AggregatedClient)
+
+	feed_id := d.Get("feed_id").(string)
+	identity_descriptor := d.Get("identity_descriptor").(string)
+	project_id := d.Get("project_id").(string)
+
+	identity_response, err := getIdentity(d, m)
+
+	if err != nil {
+		return nil, nil, err
+	}
+
+	permissions, err := clients.FeedClient.GetFeedPermissions(clients.Ctx, feed.GetFeedPermissionsArgs{
+		FeedId:  &feed_id,
+		Project: &project_id,
+	})
+
+	if err != nil {
+		return nil, identity_response, err
+	}
+
+	for _, permission := range *permissions {
+		if *permission.IdentityDescriptor == *identity_response.Descriptor {
+			return &permission, identity_response, nil
+		}
+	}
+
+	notFound := http.StatusNotFound
```

Comment:
not necessary to return the error message with 404. The create/read can check the whether `permission == nil` to check if permission is found

Reply:
Normally I would agree, but usually when 404 is thrown I see we are clearing resource : 
```
	if err != nil {
		if utils.ResponseWasNotFound(err) {
			d.SetId("")
			return nil
		}
		return fmt.Errorf("error reading feed permission during read: %+v", err)
	}
```

In the thing above, if there is no permission found during read we just clear the resource, so I can not handle it like this with simple permission == nil check, as there might be other errors which also cause permission to not exist

Reply:
@xuzhang3 any response on this one ?

Reply:
yes, we should let users decide. They can choose whether they want to restore deleted resources or create a new one.

Reply:
@xuzhang3 done, please see the last commit